UI/window-permissions.cpp

```diff
@@ -0,0 +1,101 @@
+/******************************************************************************
+    Copyright (C) 2014 by Hugh Bailey <obs.jim@gmail.com>
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 2 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+******************************************************************************/
+
+#include <QLabel>
+#include "window-permissions.hpp"
+#include "obs-app.hpp"
+
+OBSPermissions::OBSPermissions(QWidget *parent, MacPermissionStatus capture,
+			       MacPermissionStatus video,
+			       MacPermissionStatus audio,
+			       MacPermissionStatus accessibility)
+	: QDialog(parent), ui(new Ui::OBSPermissions)
+{
+	ui->setupUi(this);
+	SetStatus(ui->capturePermissionButton, capture,
+		  QTStr("MacPermissions.Item.ScreenRecording"));
+	SetStatus(ui->videoPermissionButton, video,
+		  QTStr("MacPermissions.Item.Camera"));
+	SetStatus(ui->audioPermissionButton, audio,
+		  QTStr("MacPermissions.Item.Microphone"));
+	SetStatus(ui->accessibilityPermissionButton, accessibility,
+		  QTStr("MacPermissions.Item.Accessibility"));
+}
+
+void OBSPermissions::SetStatus(QPushButton *btn, MacPermissionStatus status,
+			       const QString &preference)
+{
+	if (status == kPermissionAuthorized) {
+		btn->setText(QTStr("MacPermissions.AccessGranted"));
+	} else if (status == kPermissionNotDetermined) {
+		btn->setText(QTStr("MacPermissions.RequestAccess"));
+	} else {
+		btn->setText(
+			QTStr("MacPermissions.OpenPreferences").arg(preference));
+	}
+	btn->setEnabled(status != kPermissionAuthorized);
+	btn->setProperty("status", status);
+}
+
+void OBSPermissions::on_capturePermissionButton_clicked()
+{
+	OpenMacOSPrivacyPreferences("ScreenCapture");
+	RequestPermission(kScreenCapture);
```

Comment:
This goes for all the buttons: After the user opens the relevant System dialog and enables permissions, will the status in the dialog dynamically change?

I think the only way to do that properly would be to start a timer loop after they click a button, and keep checking until they either close the dialog, close OBS, or successfully change the permission.

Reply:
It updates dynamically only when they request the Microphone and Camera with the prompt.
However this is fine because permissions changed in the settings require an app restart to apply anyways. For example when I enable Screen Recording, I get the following popup:
<img width="307" alt="image" src="https://user-images.githubusercontent.com/59806498/181035020-bab0b112-1b16-4f60-a17a-bc014a034b0d.png">
"OBS.app cannot record the contents of your screen until it quit. You can decide whether you want to restart now or quit yourself later." (and then the buttons are "Quit & Reopen" / "Later").

What I did just notice is that the "Quit & Reopen" button doesn't actually quit OBS, which is interesting. Might need to look at what's going on there.

Reply:
In tests I've done before, that dialog isn't entirely true - I think as long as you don't attempt to access the thing before the user has allowed it, a restart is not strictly necessary. This could potentially depend on which permission it is and maybe even the macOS version.